urllib3/connection.py

```diff
@@ -162,6 +162,34 @@ def connect(self):
         conn = self._new_conn()
         self._prepare_conn(conn)
 
+    def request_chunked(self, method, url, body=None, headers={}):
+        """
+        Alternative to the common request method, which sends the
+        body with chunked encoding and not as one block
+        """
+        header_names = set(k.lower() for k in headers)
+        self.putrequest(method, url, skip_accept_encoding=True)
+        for header, value in headers.items():
+            self.putheader(header, value)
+        if 'transfer-encoding' not in header_names:
+            self.putheader('Transfer-Encoding', 'chunked')
+        self.endheaders()
+
+        if body is not None:
+            if isinstance(body, six.string_types):
+                body = [body]
+            for chunk in body:
+                if not chunk:
+                    continue
+                len_str = hex(len(chunk))[2:]
+                self.send(len_str.encode('utf-8'))
+                self.send(b'\r\n')
+                self.send(chunk.encode('utf-8'))
+                self.send(b'\r\n')
```

Comment:
So this code is still hurting in one of two cases:

- `chunk` is a Unicode string containing characters outside the ASCII charset (e.g. `u"終あ権7載チ"`). We'll want a test for this because it's easy to get wrong, but your tests actually make the same mistake! I've put a note in that should help the tests catch the error that the code is making here: if you make that change and then add my sample string to the test case you should discover the problem on the Python 3 runs.
- `chunk` is a bytestring on Python 3. On Python 3, calling `encode` on a bytestring is an error.
- `chunk` is a bytestring on Python 2 containing characters encoded using an encoding that is *not* the system default. Calling `.encode()` on a bytestring in Python 2 implicitly calls `decode()` using the system default encoding. This could easily explode.

Again, the simplest fix is to simply not allow unicode strings as chunks *at all*. If you're determined to allow them, then you should check for them explicitly at the top of the loop and encode them there, before calling `len` on them.